Added markers show conflicts of the sessions in My Schedule.

jmatsu opened the pull request:
Implemented #101 

Changes:

+ Add a conflict tag on a CardView
+ Add some resources for conflict tags
+ Refresh contents of RecyclerView when the statuses of sessions are changed

<img src="https://raw.githubusercontent.com/jmatsu/LightResources/d0eb0d96cec26f487d14caebaf3e0ba10225dfc4/reports/droidkaigi/Screenshot_2016-02-13-19-20-45.png" width="300px" alt="conflict_on_jp">

<img src="https://raw.githubusercontent.com/jmatsu/LightResources/d0eb0d96cec26f487d14caebaf3e0ba10225dfc4/reports/droidkaigi/Screenshot_2016-02-13-19-20-33.png" width="300px" alt="conflict_on_en">

konifar commented:
:eyes:

konifar commented on app/src/main/java/io/github/droidkaigi/confsched/fragment/SessionsTabFragment.java:
```diff
@@ -37,8 +40,16 @@
 
+    @Retention(RetentionPolicy.SOURCE)
+    @IntDef({ALL, MINE})
+    public @interface RENDER_MODE {}
```
I prefer enum. But I'm not sure which is better. How do you think?

hotchemi replied:
We love enum :heart:

jmatsu replied:
Sure!

konifar commented on app/src/main/java/io/github/droidkaigi/confsched/fragment/SessionsFragment.java:
```diff
@@ -185,8 +185,8 @@ protected void groupByDateSessions(List<Session> sessions) {
-    private void addFragment(String title, List<Session> sessions) {
-        SessionsTabFragment fragment = SessionsTabFragment.newInstance(sessions);
+    private void addFragment(String title, List<Session> sessions, @SessionsTabFragment.RENDER_MODE int mode) {
+        SessionsTabFragment fragment = SessionsTabFragment.newInstance(sessions, mode);
```
One suggestion
I guess it is better to change this method to `protected`, and override each chilren class. Because RENDER_MODE is only used here. What do you think?

konifar commented on app/src/main/java/io/github/droidkaigi/confsched/fragment/SessionsTabFragment.java:
```diff
@@ -137,15 +162,36 @@ public void onBindViewHolder(BindingHolder<ItemSessionBinding> holder, int posit
+                    conflictWithNext = session.stime.getTime() == nextSession.stime.getTime();
+                }
+
+                if (hideStartTime || conflictWithNext) {
```
Question.
Is this condition correct? I guess it is correct, right?

```java
if (!hideStartTime && conflictWithNext) {
```

jmatsu replied:
In this case, `session` conflict with previous session if `hideStartTime` is `true`.
`hideStartTime` means that `session` and previous session have the same `stime` (start time).

konifar commented on app/src/main/java/io/github/droidkaigi/confsched/fragment/SessionsTabFragment.java:
```diff
@@ -162,6 +208,12 @@ public void unLiked(LikeButton likeButton) {
+
+                    if (mode == MINE) {
+                        sessions.remove(session);
+                        adapter.reset(sessions);
```
Suggestion!
I guess it is better not to remove immediately. Bacause user might do miss tap. 
But I want to remove conflict tag immediately. What do you think?

konifar commented on app/src/main/res/layout/item_session.xml:
```diff
@@ -46,10 +46,25 @@
                     android:layout_width="match_parent"
                     android:layout_height="wrap_content"
+                    android:layout_below="@id/txt_conflict"
+                    android:layout_marginTop="@dimen/spacing_xsmall"
```
I guess `spacing_small` is better.

hotchemi commented on app/src/main/java/io/github/droidkaigi/confsched/fragment/SessionsTabFragment.java:
```diff
@@ -101,7 +115,19 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
+                                if (!session.checked) {
+                                    sessions.remove(session);
+                                    adapter.reset(sessions);
+                                } // else -> undefined
```
Is this commnet needed?

konifar commented:
Could you merge or rebase master branch?

jmatsu commented:
All of these reviews were reflected.

+ Use Enum instead of IntDef
+ Instantiate SessionsTabFragment with using Override
+ Remove conflict tags immediately without removing unliked sessions
+ Changed from spacing_xsmall to spacing_small
+ Removed unreasonable comments
+ Merged master-branch and reordered commits

jmatsu commented:
And, I changed how to check whether conflicts happen or not.

konifar commented:
:eyes:

konifar commented on app/src/main/java/io/github/droidkaigi/confsched/fragment/SessionsTabFragment.java:
```diff
@@ -48,13 +63,15 @@
+    public static SessionsTabFragment newInstance(List<Session> sessions, @NonNull RENDER_MODE mode) {
         SessionsTabFragment fragment = new SessionsTabFragment();
         Bundle args = new Bundle();
+        args.putInt(ARG_RENDER_MODE, mode.ordinal());
```
I guess `putSerializable()` is better to handle enum.

```java
args.putSerializable(ARG_RENDER_MODE, mode);
```

```java
mode = (RENDER_MODE) getArguments().getSerializable(ARG_RENDER_MODE);
```

jmatsu replied:
I see. :+1:

konifar commented:
Sorry for late review.
Because I think about `getRangeOfConflictCandidates()` method. I want to think more and review it. So please wait until tomorrow 1pm.

hkurokawa commented:
@jmatsu 
Hi, I am sorry for the delay in my comments.

Maybe it might sound silly to you, I think it seems not straightforward that we have `SessionsTabFragment` with a mode - ALL_SESSIONS and MY_SCHEDULE.

I feel like we can prepare two Adapters for All Sessions and My Schedule screen respectively because the information shown on the screen is different.

If you have any concerns about that, please let me know.  And if you don't mind, I will put some changes to your branch to verify how my idea goes well.

Another comment is that I am wondering you might not need to compute the possibly affected items on the list precisely. I mean, there wouldn't be so many items (right? It is just a schedule for a user) and we can just render all the items instead of introducing a complicated computation just for simplicity.

Hope it will be of your help. Thanks.

KeithYokoma commented:
Most ideally `SessionsTabFragment` should be separated into 2 fragments for each interests, but it's going to be a huge modification so I think separating adapter is the way to go.
Along with this, I think it's great to have `Model` for those 2 interests(all sessions, my schedules) so that we can easily decouple and test business logics.

jmatsu commented:
Thank you for your reviews! I agree with your ideas.

I will try to handle two current roles (all sessions and my schedule) by using each adapter.  

@hkurokawa  
You are right. There is no problem with rendering all items in RecyclerView.

In this case, `recomputing` is needless. So, I will reuse the computation results.  
I think that it is better than the current logic and rendering all items.

hkurokawa commented:
@jmatsu Thank you for your clarification and appreciate your effective support.

Just to confirm, could you let me know what "Rerendering all items causes misaligning from the current displayed position"? I guess you are saying about the current position is reset to another place when the user removes a session from My Schedule. Is my understanding correct?

jmatsu commented:
@hkurokawa I didn't mean that. Sorry.

The meaning of `the current displayed position` is `the position of the first visible item` and/or `the offset from the top of the display`.

If any conflict tags are added/removed, `the current displayed position` will be changed. This is expected and correct behavior.
However, even though any conflict tags are *not* added/removed, it sometimes happens that RecyclerView fails to keep `the current displayed position`.

*Note*
Don't disappear any session for the usability if it is unliked. cf. https://github.com/konifar/droidkaigi2016/pull/180#discussion_r52826572

hkurokawa commented:
@jmatsu Thank you for your clarification. I got it and agree with you on avoiding `the current displayed position` is changed even when any conflict tags are not added/removed. We should avoid such weird behaviour.

Let me know when you need my review.

Again, really appreciate your efficient works.  Thank you.

jmatsu commented:
I have one question.

First, the states in master-branch are followings:

+ The current adapter for sessions is an inner class of SessionsTabFragment.
+ SessionsFragment and MyScheduleFragment must handle the adapter via SessionsTabFragment.

#### Which way do you think is better? :smile:

##### One way is using Enum. Its changes would *not* be huge.
*Create new adapter in current SessionsTabFragment, then instantiate an appropriate adapter based on given Enum*

##### Another way is *not* using Enum but its changes would be huge.
*Create new fragment class (SessionsTabFragment for My Schedule) and declare new adapter in its scope.*

hkurokawa commented:
My preference is the latter - creating another Fragment for My Schedule. I would just like to avoid switching the behaviours on the mode as it would introduce a bug and some amount of cost to maintain in future.

@konifar How do you think? I would like to respect the repository owner's preference.

hkurokawa commented:
@jmatsu Hi, do you have any estimation on how it will take? We are just considering whether we can include this in the next release.

jmatsu commented:
@hkurokawa It won't be long. I'm going to prepare it soon.

konifar commented:
@jmatsu Thanks!

jmatsu commented:
Now pushed with rebasing.

hkurokawa commented:
:eyes:

hkurokawa commented on app/src/main/res/values/strings.xml:
```diff
@@ -101,6 +101,7 @@
 
     <string name="my_schedule_empty">No items here. Let\'s check out all sessions and put your favourite!</string>
     <string name="my_schedule_go_to_session_list">Go to All Sessions</string>
+    <string name="my_schedule_conflicted">Conflict happens</string>
```
This label sounds a bit weird to me. Could you change it to "Conflicted with another session" or something?

hkurokawa commented:
@jmatsu Seems good to me. Appreciate your efficient support. I added a comment on an English label. Can you update it?

hotchemi commented on app/src/main/java/io/github/droidkaigi/confsched/fragment/MyScheduleSessionsTabFragment.java:
```diff
@@ -0,0 +1,147 @@
+
+    @Override
+    protected SessionsAdapter createAdapter() {
+        return new MyScheduleAdapter(getActivity());
```
You can use getContext

jmatsu commented:
@konifar @hkurokawa @hotchemi
Updated it. And I also updated an Arabic label with using google translation.

hkurokawa commented:
@jmatsu Thanks! I am merging it.

jmatsu commented:
Thank you so much for the polite reviews! :+1:
